CORE-10466: Fix flakey upgrade vNode test

Tom-Fitzpatrick opened the pull request:
It looks like a race condition is occurring sometimes when liquibase operations are taking place during smoketests. This PR adds a call to close LiquibaseSchemaMigratorImpl after a migration or rollback is attempted. 



Below is the error seen in the db-worker logs. This seems to be causing the flakey test when we're setting the state of a Virtual Node to be Active after having just set it to be Inactive and upgrading the CPI.

```
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 11:07:17.556 [durable processing thread virtual.node.async.operation.group-virtual.node.async.request] WARN  net.corda.virtualnode.write.db.impl.writer.asyncoperation.utility.MigrationUtilityImpl {} - CPI migrations failed for virtual node '58F17573DDB8`. Failure occurred running CPI migrations on CPK with file checksum SHA-256:59EB14ADB1C043DB91D4FF6F69A1E48522B4AB23A1815F8FBDE891EEA42A98C8.
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] liquibase.exception.LiquibaseException: java.lang.RuntimeException: Cannot end scope nizxmqxfmn when currently at scope gketxjdniy
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at liquibase.Liquibase.runInScope(Liquibase.java:2450) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at liquibase.Liquibase.update(Liquibase.java:235) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at liquibase.Liquibase.update(Liquibase.java:220) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at liquibase.Liquibase.update(Liquibase.java:206) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at net.corda.db.admin.impl.LiquibaseSchemaMigratorImpl.process(LiquibaseSchemaMigratorImpl.kt:109) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at net.corda.db.admin.impl.LiquibaseSchemaMigratorImpl.updateDb(LiquibaseSchemaMigratorImpl.kt:44) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at net.corda.db.admin.impl.LiquibaseSchemaMigratorImpl.updateDb(LiquibaseSchemaMigratorImpl.kt:40) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at net.corda.virtualnode.write.db.impl.writer.asyncoperation.utility.MigrationUtilityImpl.runCpkMigrations(MigrationUtilityImpl.kt:68) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at net.corda.virtualnode.write.db.impl.writer.asyncoperation.utility.MigrationUtilityImpl.runVaultMigrations(MigrationUtilityImpl.kt:33) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at net.corda.virtualnode.write.db.impl.writer.asyncoperation.handlers.VirtualNodeUpgradeOperationHandler.tryRunningMigrationsInProcess(VirtualNodeUpgradeOperationHandler.kt:220) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at net.corda.virtualnode.write.db.impl.writer.asyncoperation.handlers.VirtualNodeUpgradeOperationHandler.upgradeVirtualNodeCpi(VirtualNodeUpgradeOperationHandler.kt:138) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at net.corda.virtualnode.write.db.impl.writer.asyncoperation.handlers.VirtualNodeUpgradeOperationHandler.handle(VirtualNodeUpgradeOperationHandler.kt:55) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at net.corda.virtualnode.write.db.impl.writer.asyncoperation.handlers.VirtualNodeUpgradeOperationHandler.handle(VirtualNodeUpgradeOperationHandler.kt:31) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at net.corda.virtualnode.write.db.impl.writer.asyncoperation.VirtualNodeAsyncOperationProcessor.onNext(VirtualNodeAsyncOperationProcessor.kt:23) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at net.corda.messaging.subscription.DurableSubscriptionImpl$ForwardingEventLogProcessor.onNext(DurableSubscriptionImpl.kt:68) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at net.corda.messaging.subscription.EventLogSubscriptionImpl.processDurableRecords$lambda$6(EventLogSubscriptionImpl.kt:231) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at io.micrometer.core.instrument.composite.CompositeTimer.recordCallable(CompositeTimer.java:129) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at net.corda.messaging.subscription.EventLogSubscriptionImpl.processDurableRecords(EventLogSubscriptionImpl.kt:231) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at net.corda.messaging.subscription.EventLogSubscriptionImpl.pollAndProcessRecords(EventLogSubscriptionImpl.kt:163) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at net.corda.messaging.subscription.EventLogSubscriptionImpl.runConsumeLoop(EventLogSubscriptionImpl.kt:126) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at net.corda.messaging.subscription.EventLogSubscriptionImpl.access$runConsumeLoop(EventLogSubscriptionImpl.kt:49) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at net.corda.messaging.subscription.EventLogSubscriptionImpl$threadLooper$1.invoke(EventLogSubscriptionImpl.kt:62) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at net.corda.messaging.subscription.EventLogSubscriptionImpl$threadLooper$1.invoke(EventLogSubscriptionImpl.kt:62) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at net.corda.messaging.subscription.ThreadLooper.runConsumeLoop(ThreadLooper.kt:174) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at net.corda.messaging.subscription.ThreadLooper.access$runConsumeLoop(ThreadLooper.kt:27) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at net.corda.messaging.subscription.ThreadLooper$StoppableThread$start$1$1.invoke(ThreadLooper.kt:106) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at net.corda.messaging.subscription.ThreadLooper$StoppableThread$start$1$1.invoke(ThreadLooper.kt:106) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at kotlin.concurrent.ThreadsKt$thread$thread$1.run(Thread.kt:30) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] Caused by: java.lang.RuntimeException: Cannot end scope nizxmqxfmn when currently at scope gketxjdniy
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at liquibase.Scope.exit(Scope.java:230) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at liquibase.Scope.child(Scope.java:191) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at liquibase.Scope.child(Scope.java:179) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at liquibase.Scope.child(Scope.java:158) ~[?:?]
[pod/corda-bob-db-worker-6844c87477-wcc8f/corda-bob-db-worker] 	at liquibase.Liquibase.runInScope(Liquibase.java:2445) ~[?:?]
```

https://r3-cev.atlassian.net/browse/CORE-10466

Tom-Fitzpatrick committed:
throw error if state update failed

Tom-Fitzpatrick committed:
fix detekt issue

Tom-Fitzpatrick committed:
close LiquibaseSchemaMigratorImpl after process/rollback

vkolomeyko commented on libs/db/db-admin-impl/src/main/kotlin/net/corda/db/admin/impl/LiquibaseSchemaMigratorImpl.kt:
```diff
@@ -34,6 +34,7 @@ class LiquibaseSchemaMigratorImpl(
         // NOTE: may need to become variable depending on the DB type
         const val DEFAULT_DB_SCHEMA = "PUBLIC"
         private val log = LoggerFactory.getLogger(this::class.java.enclosingClass)
+        private val liquibaseAccessLock = Any()
```
If locking is needed, probably best to use `ReentrantLock`.

However, this is sort of locking is only going to work inside same JVM, what is going to happen if we run multiple instances of DBWorkers? Possibly, LB has a feature of performing the lock operation on the DB level? (like: `liquibase.Liquibase#listLocks`)

owenstanford replied:
the Liquibase scope management does not appear to be thread safe, it's not a DB level issue we just need to synchronize Liquibase API access.

vkolomeyko replied:
> the Liquibase scope management does not appear to be thread safe

Sure, if this is what we are trying to address here, let's switch to use `ReentrantLock` and I will be happy to approve.

vkolomeyko commented:
Also Detekt reporting validation failure as follows:
```
/var/jenkins/workspace/Corda5_corda-runtime-os_PR-3171/libs/db/db-admin-impl/src/main/kotlin/net/corda/db/admin/impl/LiquibaseSchemaMigratorImpl.kt:110:17: Line detected, which is longer than the defined maximum line length in the code style. [MaxLineLength]	
```

Tom-Fitzpatrick committed:
fix detekt issue

vkolomeyko commented:
We probably need to understand a bit more why race condition is happening.
Perhaps, the fix could be to re-structure the Smoke test to be more patient. I.e. wait when "setting the state of a Virtual Node to be Active after having just set it to be Inactive and upgrading the CPI".

owenstanford commented:
> We probably need to understand a bit more why race condition is happening. Perhaps, the fix could be to re-structure the Smoke test to be more patient. I.e. wait when "setting the state of a Virtual Node to be Active after having just set it to be Inactive and upgrading the CPI".

I suspect it's because we have a mix of Async and old RPC style rest endpoints which is allowing the tests to invoke Liquibase in parallel, once everything is async I suspect this would not be an issue as all requests will be going through a single threaded channel. However adding synchronization to the Liquibase calls is a reasonable guard given it's a couple of lines of code and Liquibase has show itself not to be thread safe.

Tom-Fitzpatrick committed:
use ReentrantLock for synchronizing liquibase API access

vkolomeyko commented on libs/db/db-admin-impl/src/main/kotlin/net/corda/db/admin/impl/LiquibaseSchemaMigratorImpl.kt:
```diff
@@ -88,32 +90,39 @@ class LiquibaseSchemaMigratorImpl(
-            lb.update(Contexts())
-        } else {
-            lb.update(Contexts(), sql)
+        synchronized(liquibaseAccessLock) {
```
```suggestion
        liquibaseAccessLock.withLock {
```

Tom-Fitzpatrick replied:
Done, thanks

vkolomeyko commented on libs/db/db-admin-impl/src/main/kotlin/net/corda/db/admin/impl/LiquibaseSchemaMigratorImpl.kt:
```diff
@@ -122,23 +131,29 @@ class LiquibaseSchemaMigratorImpl(
-        if (liquibaseSchemaName != DEFAULT_DB_SCHEMA) {
-            log.info("Setting liquibaseSchemaName to $liquibaseSchemaName")
-            database.liquibaseSchemaName = liquibaseSchemaName
+        synchronized(liquibaseAccessLock) {
```
```suggestion
        liquibaseAccessLock.withLock {
```

Tom-Fitzpatrick replied:
Ah thank you, will fix this now

Tom-Fitzpatrick committed:
use lock.withLock instead of synchronized for thread safety

vkolomeyko reviewed (dismissed):
LGTM

vkolomeyko commented on libs/db/db-admin-impl/src/main/kotlin/net/corda/db/admin/impl/LiquibaseSchemaMigratorImpl.kt:
```diff
@@ -88,32 +91,39 @@ class LiquibaseSchemaMigratorImpl(
+                }
+                log.info("${database.connection.catalog} DB schema update complete")
+            } finally {
+                lb.close()
```
Looks like `lb.close()` caused `TruancyResponderFlowTest` to fail. We need to dig deeper why this is happening.
FAO: @owenstanford (whilst @Tom-Fitzpatrick and @conalsmith-r3 are away)
also FAO: @JamesHR3

vkolomeyko replied:
Besides instead of using:
```
try {...} finally {lb.close()}
```
we can utilize:
```
lb.use {...}
```

vkolomeyko replied:
I have pushed a change which should hopefully put this PR in a better state.

Commit:
CORE-10466: Remove call to `lb.close()`

As it is tinkering with `java.sql.Connection` which was passed into the method and which we did not create/own and therefore should not be closing.

Without this change test `net.cordacon.example.rollcall.TruancyResponderFlowTest.should verify records sent against the key and persist them` will fail as we corrupt the connection inside `net.corda.simulator.runtime.persistence.DbPersistenceService`.

vkolomeyko reviewed (approved):
LGTM
